# Introduction to sounds with python

In this lesson we'll be discussing how to use python to process sounds. We will use python to read in sound files, generate sound files, manipulate sound files, and even play sound files. 

## Read in file as numpy array
First, we'll try reading in a sound file. To do this we'll use the scipy package, and within scipy we'll be using a module called "io" (which stands for input/output). From this module, we are going to call a function that will read in a .wav file:

In [1]:
from scipy.io.wavfile import read 


Now we have access to the read function. For example, if we call the help function, it will give us some information about what "read" can do:

In [2]:
help(read)


Help on function read in module scipy.io.wavfile:

read(filename, mmap=False)
    Open a WAV file.
    
    Return the sample rate (in samples/sec) and data from an LPCM WAV file.
    
    Parameters
    ----------
    filename : string or open file handle
        Input WAV file.
    mmap : bool, optional
        Whether to read data as memory-mapped (default: False).  Not compatible
        with some bit depths; see Notes.  Only to be used on real files.
    
        .. versionadded:: 0.12.0
    
    Returns
    -------
    rate : int
        Sample rate of WAV file.
    data : numpy array
        Data read from WAV file. Data-type is determined from the file;
        see Notes.  Data is 1-D for 1-channel WAV, or 2-D of shape
        (Nsamples, Nchannels) otherwise. If a file-like input without a
        C-like file descriptor (e.g., :class:`python:io.BytesIO`) is
        passed, this will not be writeable.
    
    Notes
    -----
    Common data types: [1]_
    
    ================

In [3]:
#or, better, because it will open a pop-up dialog and not clog your code:
?read

Signature: read(filename, mmap=False)
Docstring:
Open a WAV file.

Return the sample rate (in samples/sec) and data from an LPCM WAV file.

Parameters
----------
filename : string or open file handle
    Input WAV file.
mmap : bool, optional
    Whether to read data as memory-mapped (default: False).  Not compatible
    with some bit depths; see Notes.  Only to be used on real files.

    .. versionadded:: 0.12.0

Returns
-------
rate : int
    Sample rate of WAV file.
data : numpy array
    Data read from WAV file. Data-type is determined from the file;
    see Notes.  Data is 1-D for 1-channel WAV, or 2-D of shape
    (Nsamples, Nchannels) otherwise. If a file-like input without a
    C-like file descriptor (e.g., :class:`python:io.BytesIO`) is
    passed, this will not be writeable.

Notes
-----
Common data types: [1]_

=====================  ===========  ===========  =============
     WAV format            Min          Max       NumPy dtype
=====================  ===========  ====

Here we can see from the parameters that we need to pass to the `read` function (as a minimum): a string that represents the filename. Notice that it outputs two things: a sample rate, and a numpy array representing the samples of the sounds.

Note that the function will look in the same directory where your current notebook is stored for the file. If it isn't there, it will throw an error. Use the `..` notation to tell Jupyter to look upwards in the directory tree for the file or `../NameOfFolder` if the file is in a parallel folder. 

We will call the read function on a file uploaded to the "uploaded_audio" directory (copy over from Canvas) called "Flute-A4.wav"

In [5]:
(fs, x) = read("../audio/BobMarley_track1.wav")

In [7]:
x

array([-1,  2, -2, ...,  0, -1,  1], dtype=int16)

## A few basics
Now we can ask some things about this file, for example, how long is the array? We do this by calling the "size" attribute:

In [8]:
x.size

2879830

So there are 2879830 samples in the file. If we want to know how many seconds that is, we can divide by the sampling rate (96,000 samples per second):

In [9]:
x.size/fs

29.99822916666667

We will do more in a bit with numpy and indexing, but for now:

You can index a range of values with square brackets, like this:

In [10]:
x[0:100]

array([-1,  2, -2,  2, -3,  2,  0,  0,  1, -2,  3, -3,  3, -2,  1,  0, -1,
        1, -1,  2, -1,  1,  0, -1,  1, -2,  2, -2,  3, -2,  3, -2,  1, -1,
        0,  0,  0, -1,  1, -1,  2, -2,  1, -2,  1, -1,  2,  0,  2,  0,  0,
        1,  0,  1, -1,  2,  0,  0,  2, -3,  2, -1, -2,  0, -4, -1, -4,  1,
       -4,  4, -3,  3, -1,  3, -1,  3, -2,  2, -2,  2,  0,  0,  1,  0,  0,
       -1,  0, -5,  3, -8,  3, -3,  1,  6, -2,  7, -2,  3, -2, -1],
      dtype=int16)

## Basic plotting
We can also plot the sound wave to see what it looks like. We'll do that using a popular python plotting library called matplotlib, and we'll mostly be using the pyplot module. So next, we'll import the pyplot module from the library, and give it a shortcut (or alias). Then we'll call the plot function and ask it to plot our array:

However, currently coordinates in this plot have default x-values according to their order (or "index" value; i.e., 1st, 2nd, 3rd, etc.). What would make more sense is to plot the x-axis according to time. How could we do that?

We will need to explicitly pass an array of time values to the x-axis. 
We can use the `numpy` library, which is designed for working with arrays. 
Specifically, we will need to convert sample positions to positions in time.

We're going to use the "arange" function. Let's see what it does:

This is a really useful function and we'll likely be using it a lot. It creates a numpy array that begins with the "start" value, ends (before) the "stop" value, and increments by the "step" value. Notice that the "start" and "step" arguments are optional and have default values of 0 and 1, respectively. 

Notice also the line about incrementing by non-integer values; in this case it's better to use the function linspace instead of arange. You can play around with them and test it out.

(You will also want to read about `np.linspace`, which is similar and we will also use a lot)

What will be the result of this function?:

Our variable "x" is an array containing the amplitude values for every sample. Here we use `np.arange` to convert the array to essentially its index value (or position)--so far doing exactly what `pyplot` was doing by default. 
However, if we then divide each value by the sample rate, we will have an array representing each sample's position on a time axis beginning from zero.

Notice that this is a vector operation (it automatically applied the division to every item in the array). Now we have a series of time values that "line up with" our sample values. 

We can now redo our plot but this time plot *t* with respect to *x*:

We should really always add titles and label our axes, so let's do that:

We can also play our sound back and hear what it sounds like. We'll do this by using the `Audio` function from IPython. 

Audio will accept many files types including mp3, but we'll only be working with wav files for now. Note that audio also accepts a numpy array. However, you should be *extremely* cautious before playing back any numpy arrays you have created. **Always play back at very low volume first!!!**

You can also pass `Audio()` a numpy array. However, you'll need to provide a sample rate as one of the arguments (or else the function doesn't know what to do)! It needs the sample rate to be able to interpret the array. Like so: